src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -415,6 +426,11 @@ private void ShutdownWrite(Exception? shutdownReason)
 
         public override async ValueTask DisposeAsync()
         {
+            if (_stream == null)
+            {
+                return;
```

Comment:
This scares me considering DisposeAsync() pools. Does this mean something could still have a reference to the QuicStreamContext and dispose it after pooling? Maybe we should move the pooling logic outside of DisposeAsync since this is exposed to connection middleware and DisposeAsync() can be called multiple times.

Reply:
I added that Just In Case. It could be removed.

When a QUIC stream is disposed the callback removes it from the connection before it is pooled. And only gracefully completed requests are pooled. No active request will be using a pooled stream.

Reply:
What happens if connection middleware disposes the ConnectionContext before exiting?

Reply:
If the app (which is Kestrel HTTP/3 implementation in this case) hasn't completed the pipes then it isn't pooled.

https://github.com/dotnet/aspnetcore/blob/a079651ab7bfe2c086ab26567c12364f0a4f2480/src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs#L418-L422

Reply:
But third-party middleware could complete the Pipes, right? Sorry to randomize this PR. I do not think the pooling logic needs to be reworked before merging this. And I'll admit this likely isn't the most likely scenario. It's just seeing this got me thinking about pooling too early, and if it *did* become a problem, it might be very hard to diagnose.

Reply:
Yes.

There has to be some public API and condition at which point the context is pooled. Whatever we use, there is the ability for someone to use the public API to trigger pooling early.